### ISSUE TITLE
[2.3] Deadlock with Localtalk and AFP

### ISSUE THREAD

**Issue by xhero** [Netatalk/netatalk#719]
I've been testing Netatalk with my homebrew Localtalk driver for Linux which shows up as an lt0 interface. I have then and interface using the TashTalk adapter which plugs via USB and shows as a serial port on my host machine. My Netatalk configuration is extremely basic, atalkd.conf:

```
lt0 -seed -phase 2 -net 0 -addr 0.129 -zone "LocalTalk"
```

My network is like this: `PC with Netatalk -> LT2USB -> Localtalk -> PowerBook 100 via printer port`

Netatalk works quite well: I see my shares, I can print, time gets updated, I can mount shares and copy files from the PC, but not vice-versa: if I try to copy a file from the Mac to the PC share it copies the first chunk and then stops forever. The Mac then sends NBP packets to find the server again until it timeous in about 15/20 minutes. The server never responds. I tried this with my PB100 and PDQ G3.
After many hours of digging I figured out that Netatalk discards the incoming ATP response from the mac. In particular in this function: https://github.com/Netatalk/netatalk/blob/7296eb5362c6f288eba1db169a5b5935fcbd2847/libatalk/atp/atp_packet.c#L311
Basically the packet coming from the mac (the address in saddr) has s_net == ATADDR_ANYNET, but the function does not check for that, so it returns false. I fixed this adding `saddr->sat_addr.s_net == ATADDR_ANYNET` on line 320. Then it works without problems.
I'm not really sure if this is the right fix, or if it is a problem with my configuration on in the kernel routing, but this was the only way I could get AFP to work.

**Comment by NJRoadfan** [comment]
Test it with the network number set greater than "0" in atalkd.conf. Don't recall if having a network number of "0" as being valid on networks with routers and zones. I don't think my FastPath allows it for example. The Linux kernel itself has may quirks with LocalTalk networks in general (ex: node number 254 can't be used even though it is valid on LT) since it was basically hacked in.

**Comment by xhero** [comment]
So I tried to change the network to different values all > 0, I can see and mount the share but it blocks in the same place. In any case the Mac always seems to get 0 as a network number:

```
~/netatalk_local/bin$ ./nbplkup 
                        pc:ProDOS16 Image                     54321.129:3
                        pc:Apple //e Boot                     54321.129:3
                Netatalk Server:AFPServer                     54321.129:128
                        pc:Apple //gs                         54321.129:3
                        pc:TimeLord                           54321.129:132
                          PB100:AFPServer                     0.1:252
                          PB100:PowerBook 100                 0.1:253
                          PB100:Workstation                   0.1:4
```
54321.129 is the PC, 0.1 is the mac (which was started before I turned on the pc)

**Comment by xhero** [comment]
Aha! I changed from -seed to -router and now the mac gets the proper (54321) network number and it does not deadlock anymore!

**Comment by NJRoadfan** [comment]
Make sure you are running atalkd from Netatalk 2.3.1 as it fixes a critical bug regarding RTMP broadcasts. The Powerbook should be picking up the network number and zone information from ZIP packets being sent from the netatalk box.

Also, the `-seed` switch is for machines with more then one interface running AppleTalk. You have to use `-router` if you want to seed a network on a machine with only one interface.

**Comment by xhero** [comment]
ok yes I see, it was really non obvious, the documentation on how to use localtalk is 20 years old and the config syntax is not even valid anymore. Maybe I could write down a couple notes on this? BTW it even works with multiple localtalk interfaces!
In any case would it make sense to find a fix for the ATP issue? The thing here is that when it is triggered, everything just silently dies and there is really no way to figure out what is happening.

**Comment by rdmark** [comment]
Regarding what kind of fix to apply: I think we should at the very least fail early rather than stall for 15 min. Looking at the at_addr_eq function, the overall logic seems to be to check if the primary address gets a valid response, OR that the secondary address is the same as the primary address. The proposed fix seems to change the logic so that the primary address with no response or invalid response becomes a valid case, as long as the secondary address has one. I don’t know how this maps to the AppleTalk spec, but my guess is the current logic is correct, and that we should add error handling elsewhere, rather. For instance, prevent atalkd from starting when -seed is applied when only one interface is configured?

Happy to hear counter arguments! :)

I agree that we have a documentation deficit here. In fact, the manual and atalkd man page doesn’t even mention the `-router` switch: 

https://netatalk.io/2.3/htmldocs/configuration#idm789
https://netatalk.io/2.3/htmldocs/atalkd.8

The atalkd.conf man page does document it though, and does correctly call out that -seed should not be used on with a single interface:

https://netatalk.io/2.3/htmldocs/atalkd.conf.5

**Comment by xhero** [comment]
Regarding my "fix" it was kinda a bushfix and unfortunately I'm not expert enough to tell if it would have side effects. But I agree that everything should fail beforehand, it is very disconcerting that everything just... dies.
In any case I can report that using multiple lt* with `-seed` works perfectly! And traffic gets routed between the two interfaces.

I think my `-seed` came from the examples with the COPS driver which were only for multiple interfaces (and a super old syntax that does not work anymore). While now it all makes sense, I remember hitting my head on the desk quite hard trying to make sense of the docs and to figure out a configuration that would work (like what does `-router` do except for being like `-seed` and the opposite of `-dontroute`, when do I need it? What do I route if there is only one interface. Then seed does not really tell what it does!) I can submit a PR to clean up the docs a bit.

**Comment by NJRoadfan** [comment]
`-seed`: Seeds a configured network number and zone names on that interface. By default, atalkd probes a network to see if there is a router with zones on it. If one is found, it will configure that interface for the discovered network and zones. So by default it "soft seeds".
`-router`: Same as `-seed` but if you only have one interface on the machine.
`-dontroute`: This tells atalkd to not route traffic from this interface to the other interfaces on the machine. For example, if you have three interfaces named `eth0`, `eth1`, and `lt0` and use `-dontroute` on `eth1`, traffic from `eth0` and `lt0` will not be relayed to/from `eth1`.

**Comment by rdmark** [comment]
@xhero a refresh of the docs would be great!

**Comment by rdmark** [comment]
@xhero Here's an attempt to fail early in this scenario: https://github.com/Netatalk/netatalk/pull/720

Can you test if it works as expected in your environment?

**Comment by rdmark** [comment]
The fix is in the `branch-netatalk-2-3` branch now. It works as expected in my environment so I'm closing the ticket, but please don't hesitate to reopen if you find more bugs in this area.

**Comment by xhero** [comment]
@rdmark sorry I missed this, I'm traveling right now, will give it a test asap next week when I'm back, thanks!

**Comment by rdmark** [comment]
@xhero Take your time!

I'm also looking forward to your updates to the AppleTalk docs. :)
